### PR TITLE
Add 'gzip' plugin

### PR DESCRIPTION
**Pull request by into-the-v0id**
## Description

Add plugin to compress files with gzip.

Uses https://deno.land/x/foras@v2.1.4

## Related Issues

None

### Check List

- [x] Have you read the
      [CODE OF CONDUCT](https://github.com/lumeland/lume/blob/master/CODE_OF_CONDUCT.md)
- [x] Have you read the document
      [CONTRIBUTING](https://github.com/lumeland/lume/blob/master/CONTRIBUTING.md)
  - [x] One pull request per feature. If you want to do more than one thing,
        send multiple pull request.
  - [x] Write tests.
  - [x] Run deno `fmt` to fix the code format before commit.
  - [x] Document any change in the `CHANGELOG.md`.

### CODE REVIEW

File: deps/foras.ts
Comment (oscarotero) [inline deps/foras.ts:1]: Did you try this? https://docs.deno.com/api/web/~/CompressionStream
It's a web standard supported by Deno. If it outputs the same result, we don't need a dependency.

File: plugins/gzip.ts
Comment (oscarotero) [inline plugins/gzip.ts:32]: Why are you using event here?
With a processor, you don't need to import the `pageMatches` function.

```js
site.process(options.extensions, (pages) => {
  for (const page of pages) {
    page.content = compress(page.content);
    page.data.url += ".gz";
  }
});
```

**Review by oscarotero — CHANGES_REQUESTED** [review]
Thanks for this. I left a couple comments, let me know your thoughts!

Also, [I have this issue](https://github.com/lumeland/lume/issues/664) to create a middleware to serve precompressed files, so this plugin would be the first step to implement it!

File: plugins/gzip.ts
Comment (oscarotero) [inline plugins/gzip.ts:21]: Maybe the default value should be `*`, to support all files.

File: deps/foras.ts
Comment (into-the-v0id) [reply deps/foras.ts:1]: I have seen it but since you cannot pass options to it (level, etc.) I used https://deno.land/x/foras@v2.1.4 where you have more granular control of how it's compressed.

There is also `node:zlib` which also allows you to use gzip with options. However I wanted to avoid node libs since lume does not use any as of now.

However, I am absolutely fine with `CompressionStream` too. If you want me to, I would happily update the PR to use `CompressionStream`.

File: plugins/gzip.ts
Comment (into-the-v0id) [reply plugins/gzip.ts:32]: To ensure the compression happens last. If I were to use a processor instead, given the following config:

```js
site.use(gzip())
site.use(minifyHTML())
```

It would output:

- `.html` files - minified, uncompresed
- `.html.gz` files - unminified, compresed

So the compressed and uncompressed files divert. Additionally some plugins create/add pages (e.g. the sitemap plugin) which might not get compressed with a processor.

If you think that behavior is ok and want to put the responsibility on the user correctly ordering the plugins I can update the PR to use a processor.

File: plugins/gzip.ts
Comment (into-the-v0id) [reply plugins/gzip.ts:21]: I think the default here should be fine. Since `*` would also compress images, videos and other binary files. Most binary files have their own compression and gzip is not that well suited for binary/non-text data. What do you think?

File: deps/foras.ts
Comment (oscarotero) [reply deps/foras.ts:1]: I like the plugins that "just works", with minimum configuration as possible. If you think the compression level is an important thing that many people like to adjust, the dependency is okay. But if you think the CompressionStream is good enough for most cases, I'd use it.

File: plugins/gzip.ts
Comment (oscarotero) [reply plugins/gzip.ts:32]: For now, I prefer to make the order explicit in the _config.ts file because it's more transparent and is aligned with the behavior of other plugins that requires a especific order (base_path before feed, tailwindcss before postcss, to name only a few).
This also allows to place other plugins after gzip if you don't want to compress the pages that these plugins can generate.

If at some point we need to implement a way to "autosort" the plugings to be independent of the order in the _config.ts file, that solution should work for all plugins.

File: plugins/gzip.ts
Comment (oscarotero) [reply plugins/gzip.ts:21]: Fair.
Let's default to these extensions then.

File: plugins/gzip.ts
Comment (into-the-v0id) [reply plugins/gzip.ts]: All right - I have updated the PR to use a processor.

In this case we should have a look at the plugins `robots`, `sitemap` and `feed`. They use the `beforeSave` event when they could (probably) use a processor instead (plugins `esbuild` and `on_demand` use the event too but there it seems justified). Because otherwise this:

```js
site.use(sitemap())
site.use(gzip())
```

would only output a `sitemap.xml` without a compressed counterpart.

Comment (oscarotero) [reply plugins/gzip.ts]: Oh, now I see your point much better.

Yeah, `robots`, `sitemap` and `feed` should be changed to processors. But that is a breaking change.
I think maybe a good solution is to implement something like [render order](https://lume.land/docs/core/render-order/) for processors. For example:

```js
site.process([".html"], fn1, 2); // Third
site.process([".html"], fn2, 1); // Second
site.process([".html"], fn3, -1); // First
```

By default is `0` but some plugins can register processors in different order to ensure they are executed after or before others.

**Review by oscarotero — APPROVED** [review]

**Comment by oscarotero** [comment]
This is great. Thanks so much!

File: plugins/gzip.ts
Comment (into-the-v0id) [reply plugins/gzip.ts]: That would be one solution. But I think using regular processors and having the user decide on the order is still better. It would be a breaking change to update those plugins but there is no hurry to update them so it can probably wait a bit.

Using something like the render order for processors adds unnecessary complexity imo. Render order works for pages, because you have an overview about _all_ the pages so you can adjust the render order accordingly. With plugins however you dont have an overview about all plugins and their processor order. So some plugin processor orders will still be equal and require ordering from the user and others wont. Some might not even work properly anymore because a plugin needs to run before plugin A with priority 1 and after plugin B with priority 2. This would mean that maintainers of independent plugins might need to talk to other plugin maintainers to find the "right" order that works for that scenario.

Comment (oscarotero) [reply plugins/gzip.ts]: that's a fair point. 
Okay, let's keep it as is.
Thanks again1

**Pull request merged**
